### ISSUE TITLE
Consider providing the necessary semantics for denoting tab bars for accessibility

### ISSUE THREAD

**Issue by vick08** [flutter/flutter#107861]
Presently, there seems to be no way to mark up custom tab bars with accessibility semantics. As a result, developers resort to simulating them by "stuffing" the accessibility lable with announcements such as:
"Notifications, tab 1 of 3"
Etc.

iOS, Android and web provide ways to denote tab bars to match native accessibility APIs for these platforms. SOme useful resources below:
- iOS: [iOS accessibility tabBar trait](https://developer.apple.com/documentation/uikit/uiaccessibilitytraits/1648592-tabbar), [real-life implementation](https://www.basbroek.nl/custom-tab-bar-accessibility).
- Android: setRoleDescription API
- Web: [ARIA tabpannel pattern](https://www.w3.org/WAI/ARIA/apg/patterns/tabpanel/)

/CC @goderbauer

**Comment by goderbauer** [comment]
When we expose this API in Flutter we should also use it for Flutter's implementation of the Tab bar, which currently also simulates the announcement via labels.

@chunhtai Have you by any chance looked at these more specific tab bar a11y APIs before?
cc @jmagman (iOS) @GaryQian (Android) @yjbanov (web)

**Comment by chunhtai** [comment]
no I will have not looked into this before. will have to test if the mentioned APIs work as expected

**Comment by chunhtai** [comment]
@hangyujin Can you take a look at this?

**Comment by ajanec01** [comment]
That's correct. The drawback of conveying tabs in the label is that the iOS Switch Control focus indicator does not navigate to them... well, only to the selected tab as it has a "programmatically determined state". The rest of the tabs do not have a "programmatically determined role" and so switch control is treating them as text nodes.

**Comment by vick08** [comment]
1. Providing control types in the label also has the following side effects: they will be fully spelled out on the Braille display regardless of the user configuration. Tangentially, assistive technology has no way of modifying the presentation of such controls to the user since they are forced as part of the label.
2. FWIW, these APIs are used in both Material iOS and material Android components.

**Comment by jwlilly** [comment]
For Android, it seems like this may be able to be solved be adding setting a role description of "Tab" on the AccessibilityNodeInfo. The role description is part of the `AccessibilityNodeInfoCompat` wrapper class, but it can be manually added to `AccessibilityNodeInfo` using `info.getExtras().putCharSequence("AccessibilityNodeInfo.roleDescription", roleDescription);`: [AccessibilityNodInfoCompat source](https://android.googlesource.com/platform/frameworks/support/+/48beb83b59b0062adb88694e875b3e4ac8a48757/compat/src/main/java/androidx/core/view/accessibility/AccessibilityNodeInfoCompat.java). The index announcements such as "1 of 3" should come from the `CollectionInfo` and `CollectionItemInfo`. Based on my glance at the AccessibilityBridge for Android, it doesn't look like these are implemented for anything other than scroll views. I think there may be another issue out there for these two properties on Android. 

iOS is going to be a little different. It needs the `tabbar` trait on the parent view with the actual tab controls having the `button` trait. The parent view with the `tabbar` trait needs to also be hidden from assistive technology with the child views exposed. The index announcements such as "1 of 3" will be automatically added by VoiceOver.

**Comment by chunhtai** [comment]
I composed a list of necessary role we should add
http://flutter.dev/go/semantics-roles

it looks like sooner or later we would need to add tar bar role

**Comment by jcollins-disney** [comment]
@darshankawar and @chunhtai - The team at Disney prefers that these all be tracked individually we can communicate incremental progress for each missing role and their associated ARIA properties and keyboard support.

**Comment by darshankawar** [comment]
Thanks for the feedback @jcollins-disney 
I will reopen them so as to track them separately.

**Comment by Tiska** [comment]
hi [darshankawar](https://github.com/darshankawar). We are also interested by this issues. Do you have any information about the roadmap of this ? And do you know where in the Flutter repo we could make a PR to work on one of this topics ?

**Comment by Tiska** [comment]
Do you think we should add role by role or add WE could add a property to add any aria role we pass for web for example ? Could you indicate where the aria mapping is done ?

**Comment by chunhtai** [comment]
the mapping i have outlined in this doc http://flutter.dev/go/semantics-roles

The idea is to have an enum like
```dart

enum SemanticsRole {
   tabBar,
   tab,
   tabList
   list,
   listItem,
  ...
}

// And then you can do

Semantics(
  role: SemanticsRole.tabBar
)
```

**Comment by Tiska** [comment]
Ok like the button role ? And then where the mapping for the web component is done ?

**Comment by jwlilly** [comment]
> the role has added, just need to make sure each platform reads it correctly
> 
> SemanticsRole.tab
> 
> OS	role
> web	tab
> ios	-
> macos	NSAccessibilityButtonRole, NSAccessibilityTabButtonSubrole
> windows	ROLE_SYSTEM_PAGETAB
> android	-
> SemanticsRole.tabBar
> 
> OS	role
> web	tabpanel
> ios	-
> macos	-
> windows	ROLE_SYSTEM_PROPERTYPAGE
> android	-
> SemanticsRole.tabPanel
> 
> OS	role
> web	tablist
> ios	[UIAccessibilityTraitTabBar](https://developer.apple.com/documentation/uikit/uiaccessibilitytraits/tabbar?language=objc)
> macos	-
> windows	ROLE_SYSTEM_PAGETABLIST
> android	com.google.android.material.tabs.TabLayout.TabView (?)

@chunhtai is there any way I can help you test this out? The only concern I have is with Android. I'm not sure how you implemented it in the backend. If you extended from that com.google.android.material.tabs.TabLayout.TabView, then it might work, but the tab role is implemented (even by that class), using [AccessibilityNodeInfoCompat.setRoleDescription("Tab")](https://developer.android.com/reference/androidx/core/view/accessibility/AccessibilityNodeInfoCompat#setRoleDescription(java.lang.CharSequence)).

**Comment by chunhtai** [comment]
somewhere in this method you can set class name like this line https://github.com/flutter/flutter/blob/0369b3564043a0b68a143bac68dd265545d2d2c9/engine/src/flutter/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L675 for a certain node id (by checking the virtualViewId)

and then see if talkback treats it differently

setRoleDescription may not be desired since then we will have to own the localization

**Comment by jwlilly** [comment]
> somewhere in this method you can set class name like this line
> 
> [flutter/engine/src/flutter/shell/platform/android/io/flutter/view/AccessibilityBridge.java](https://github.com/flutter/flutter/blob/0369b3564043a0b68a143bac68dd265545d2d2c9/engine/src/flutter/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L675)
> 
> Line 675 in [0369b35](/flutter/flutter/commit/0369b3564043a0b68a143bac68dd265545d2d2c9)
> 
>  result.setClassName("android.view.View"); 
> for a certain node id (by checking the virtualViewId)
> and then see if talkback treats it differently
> 
> setRoleDescription may not be desired since then we will have to own the localization

I think using setClassName will resolve to "LinearLayout" or just a "View" and no role will be announced by TalkBack. A tab role in Android doesn't really exist based on className. It really only works with the base widgets like button, checkbox, radio button, etc. A roleDescription is how Google is setting it in the Play Store: 

![Screenshot of accessibility inspector showing Google Play store tabs](https://github.com/user-attachments/assets/a24a8e68-78c8-4af8-8bf0-e7f84c3e07db)

And in the source of "TabView": https://github.com/material-components/material-components-android/blob/d4ac3d593e6d341c88acea0aa8cf6dda8fa2afed/lib/java/com/google/android/material/tabs/TabLayout.java#L2664

**Comment by chunhtai** [comment]
looks like https://api.flutter.dev/flutter/material/NavigationBar-class.html is missing tabbar role

**Comment by chunhtai** [comment]
current status, the web engine part is already done, the ios and android is still missing, but not sure whether this is still blocking the customer

**Comment by kevmoo** [comment]
@chunhtai – is there more to do on mobile, or are we as done as we can be?

**Comment by chunhtai** [comment]
both ios and Android parts are missing.

**Comment by chunhtai** [comment]
Web part of this is done.

**Comment by kevmoo** [comment]
@chunhtai is this fixed for web or are we still waiting on https://github.com/flutter/flutter/pull/162467 ?
